src/core/abr/learn_2_adapt.ts

```diff
@@ -0,0 +1,260 @@
+
+  constructor({ bitrates }: ICxtArg) {
+    // this._representations = representations;
+    this._bitrates = bitrates;
+
+    this._state = L2A_State.Start;
+    this._lastQuality = 0;
+    this._lastSegmentDuration = 0;
+    // this._placeholderBuffer = 0;
+    // this._lastSegmentWasReplacement = false;
+    this._mostAdvancedSegmentStart = null;
+    this._lastSegmentDuration = null;
+    this._lastSegmentStart = null;
+    this._lastSegmentRequestTimeMs = null;
+    this._lastSegmentFinishTimeMs = null;
+
+    this._l2AParameters = {
+      w: [],
+      prevW: [],
+      Q: 0,
+      segmentRequestStart: 0,
+      segmentDownloadFinish: 0,
+      bufferTargetLvl: 1.5,
+    };
+  }
+
+  private _resetL2AInfos() {
+    // this._placeholderBuffer = 0;
+    // this._lastSegmentWasReplacement = false; // used to know if the last segment has been replace (we dont use it for now);
+    this._mostAdvancedSegmentStart = null;
+    this._lastSegmentDuration = null;
+    this._lastSegmentStart = null;
+    this._lastSegmentRequestTimeMs = null;
+    this._lastSegmentFinishTimeMs = null;
+  }
+
+  /**
+   * Adjust State accordingly
+   */
+  private _checkNewSegment() {
+    if (this._lastSegmentStart !== null &&
+        this._lastSegmentRequestTimeMs !== null &&
+        this._lastSegmentFinishTimeMs !== null) {
+      this._l2AParameters.segmentRequestStart = 0.001 * this._lastSegmentRequestTimeMs;
+      this._l2AParameters.segmentDownloadFinish = 0.001 * this._lastSegmentFinishTimeMs;
+      this._lastSegmentStart = null;
+      this._lastSegmentRequestTimeMs = null;
+    }
+  }
+
+  private _dotmultiplication(a: number[], b: number[]) {
+    if (a.length !== b.length) {
+      return -1;
+    }
+    let sumdot = 0;
+    for (let i = 0; i < a.length; i++) {
+      sumdot = sumdot + a[i] * b[i];
+    }
+    return sumdot;
+  }
+
+  /**
+     * Project an n-dim vector y to the simplex Dn
+     * Dn = { x : x n-dim, 1 >= x >= 0, sum(x) = 1}
+     * Algorithm is explained at http://arxiv.org/abs/1101.6081
+     * @param {array} arr
+     * @return {array}
+     */
+  private _euclideanProjection(arr: number[]): number[] {
+    const m = arr.length;
+    let bget = false;
+    let arr2 = [];
+    for (let ii = 0; ii < m; ++ii) {
+        arr2[ii] = arr[ii];
+    }
+    let s = arr.sort(function (a, b) {
+        return b - a;
+    });
+    let tmpsum = 0;
+    let tmax = 0;
+    let x = [];
+    for (let ii = 0; ii < m - 1; ++ii) {
+        tmpsum = tmpsum + s[ii];
+        tmax = (tmpsum - 1) / (ii + 1);
+        if (tmax >= s[ii + 1]) {
+            bget = true;
+            break;
+        }
+    }
+    if (!bget) {
+        tmax = (tmpsum + s[m - 1] - 1) / m;
+    }
+    for (let ii = 0; ii < m; ++ii) {
+        x[ii] = Math.max(arr2[ii] - tmax, 0);
+    }
+    return x;
+}
+
+  /**
+   * When the Player seek, reset the state to Start and
+   * reset L2A infos, to start from a fresh learning copy.
+   */
+  public onSeeking() {
```

Comment:
Not yet listened to?

I'm not sure what exact case it should trigger at but we could just be based on the clock's `seeking` and `internal-seeking` events.